packages/utils/src/promise.ts

```diff
@@ -25,115 +27,136 @@ export async function callFnWhenAwait<T>(
-      .catch((e: Error) => {
-        eventCb(RaceEvent.rejected, Date.now() - startTime, index);
-        promisesStates[index] = {status: PromiseStatus.rejected, value: e};
-      });
-  });
-
-  // Wait till cutoff time unless all original promises resolve/reject early
-  await Promise.allSettled(promises.map((promise) => Promise.race([promise, cutoffPromise])));
-  if (cutoffObserved) {
-    // If any is resolved, then just simply return as we are post cutoff
-    const anyResolved = promisesStates.reduce(
-      (acc, pmState) => acc || pmState.status === PromiseStatus.resolved,
-      false
-    );
-    if (anyResolved) {
-      eventCb(RaceEvent.resolvedatcutoff, Date.now() - startTime);
-      return mapStatusesToResponses(promisesStates);
-    } else {
-      eventCb(RaceEvent.cutoff, Date.now() - startTime);
-    }
-  } else {
-    eventCb(RaceEvent.precutoff, Date.now() - startTime);
-    return mapStatusesToResponses(promisesStates);
+export async function resolveOrRacePromises<T extends NonEmptyArray<Promise<unknown> | PromiseWithStatus<unknown>>>(
+  promises: T,
+  {
+    resolveTimeoutMs,
+    raceTimeoutMs,
+    signal,
+  }: {
+    resolveTimeoutMs: number;
+    raceTimeoutMs: number;
+    signal?: AbortSignal;
+  }
+): Promise<ReturnPromiseWithTuple<T>> | never {
+  if (raceTimeoutMs <= resolveTimeoutMs) {
+    throw new Error("Race time mus tbe greater than resolve time");
   }
 
-  // Post deadline resolve with any of the promise or all rejected before timeout
-  await Promise.any(promises.map((promise) => Promise.race([promise, timeoutPromise]))).catch(
-    // just ignore if all reject as we will returned mapped rejections
-    // eslint-disable-next-line  @typescript-eslint/no-empty-function
-    (_e) => {}
+  const promisesWithStatuses = promises.map((p) => (p instanceof PromiseWithStatus ? p : new PromiseWithStatus(p)));
+  const resolveTimeoutError = new TimeoutError(
+    `Given promises can't be resolved within resolveTimeoutMs=${resolveTimeoutMs}`
+  );
+  const raceTimeoutError = new TimeoutError(
+    `Not a any single promise be resolved in given raceTimeoutMs=${raceTimeoutMs}`
   );
-  if (timeoutObserved) {
-    eventCb(RaceEvent.timeout, Date.now() - startTime);
-  } else {
-    eventCb(RaceEvent.pretimeout, Date.now() - startTime);
+
+  try {
+    await Promise.race([
+      Promise.allSettled(promisesWithStatuses),
+      sleep(resolveTimeoutMs, signal).then(() => {
+        throw resolveTimeoutError;
+      }),
+    ]);
+    return promisesWithStatuses as ReturnPromiseWithTuple<T>;
+  } catch (err) {
+    if (err !== resolveTimeoutError) {
+      throw err;
+    }
   }
-  return mapStatusesToResponses(promisesStates);
+
+  try {
+    await Promise.race([
+      Promise.any(promisesWithStatuses),
+      sleep(raceTimeoutMs - resolveTimeoutMs, signal).then(() => {
+        throw raceTimeoutError;
+      }),
+    ]);
+  } catch (err) {
+    if (err !== raceTimeoutError && !(err instanceof AggregateError)) {
```

Comment:
since promisesWithStatues doesn't throw, AggregateError generated from Promise.any means some internal error and should throw so the condition should just be `err!==raceTimeoutError`

Reply:
I believe AggregateError here means that all promises passed to Promise.any have rejected.
I think current implementation is correct because it lets us handle this with the promises with statuses rather than in a catch.

Reply:
correct but promisesWithSttaues can't fail unless there is an internal error as the rejection just updates the state, so AggregateError couldn't be thrown anyway